File: backend/app/api/graph.py
"""
Graph query API routes.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status, Query

from app.api.dependencies import get_graph_query_service
from app.graph.exceptions import GraphEntityNotFoundError
from app.graph.models import GraphNode
from app.graph.query_service import GraphQueryService
from app.graph.query_models import GraphNeighborsResponse
from app.graph.query_models import (
    GraphNeighborsResponse,
    RelatedIncidentsResponse,
    TopRiskEntityResponse,
)
from app.graph.query_models import EntityRiskResponse
from app.graph.query_models import FraudRingResponse
from app.graph.query_models import NetworkSummaryResponse
from app.graph.query_models import PathResponse
from app.graph.query_models import SharedEntityResponse

router = APIRouter()


@router.get(
    "/entity/{value}",
    response_model=GraphNode,
    summary="Retrieve a graph entity",
)
async def get_entity(
    value: str,
    service: Annotated[
        GraphQueryService,
        Depends(get_graph_query_service),
    ],
) -> GraphNode:
    """
    Retrieve a graph entity by its value.
    """
    try:
        return await service.get_entity(value)

    except GraphEntityNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

@router.get(
    "/entity/{value}/neighbors",
    response_model=GraphNeighborsResponse,
    summary="Retrieve neighboring graph entities",
)
async def get_neighbors(
    value: str,
    service: Annotated[
        GraphQueryService,
        Depends(get_graph_query_service),
    ],
) -> GraphNeighborsResponse:
    """
    Retrieve an entity together with its directly connected neighbors.
    """
    try:
        return await service.get_neighbors(value)

    except GraphEntityNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

@router.get(
    "/entity/{value}/incidents",
    response_model=RelatedIncidentsResponse,
    summary="Retrieve related incidents",
)
async def get_related_incidents(
    value: str,
    service: Annotated[
        GraphQueryService,
        Depends(get_graph_query_service),
    ],
) -> RelatedIncidentsResponse:
    """
    Retrieve all complaint nodes connected to an entity.
    """
    try:
        return await service.get_related_incidents(value)

    except GraphEntityNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(exc),
        ) from exc

@router.get(
    "/entity/{value}/risk",
    response_model=EntityRiskResponse,
)
async def get_entity_risk(
    value: str,
    service: GraphQueryService = Depends(get_graph_query_service),
) -> EntityRiskResponse:
    """
    Retrieve risk assessment for a graph entity.
    """
    return await service.get_entity_risk(value)

@router.get(
    "/entity/{value}/ring",
    response_model=FraudRingResponse,
)
async def get_fraud_ring(
    value: str,
    service: GraphQueryService = Depends(
        get_graph_query_service,
    ),
) -> FraudRingResponse:
    """
    Retrieve the connected fraud ring for an entity.
    """
    return await service.get_fraud_ring(value)

@router.get(
    "/network/summary",
    response_model=NetworkSummaryResponse,
)
async def get_network_summary(
    service: GraphQueryService = Depends(
        get_graph_query_service,
    ),
) -> NetworkSummaryResponse:
    """
    Retrieve graph-wide statistics.
    """

    return await service.get_network_summary()

@router.get(
    "/network/top-risk",
    response_model=list[TopRiskEntityResponse],
)
async def get_top_risk_entities(
    limit: int = Query(
        default=10,
        ge=1,
        le=100,
    ),
    service: GraphQueryService = Depends(
        get_graph_query_service,
    ),
) -> list[TopRiskEntityResponse]:
    """
    Retrieve the highest-risk entities in the graph.
    """

    return await service.get_top_risk_entities(limit)

@router.get(
    "/path",
    response_model=PathResponse,
)
async def get_shortest_path(
    source: str,
    target: str,
    service: GraphQueryService = Depends(
        get_graph_query_service,
    ),
) -> PathResponse:
    """
    Find the shortest path between two graph entities.
    """

    return await service.get_shortest_path(
        source,
        target,
    )

@router.get(
    "/entity/{value}/shared",
    response_model=SharedEntityResponse,
)
async def get_shared_entity(
    value: str,
    service: GraphQueryService = Depends(
        get_graph_query_service,
    ),
) -> SharedEntityResponse:
    """
    Find all complaints sharing the specified entity.
    """

    return await service.get_shared_entity(
        value,
    )

File: frontend/src/api/graph.ts
import { apiFetch } from './client';
import {
  EntityRiskResponse,
  FraudRingResponse,
  GraphNeighborsResponse,
  GraphNode,
  GraphResponse,
  NetworkSummaryResponse,
  PathResponse,
  RelatedIncidentsResponse,
  SharedEntityResponse,
  TopRiskEntityResponse,
} from '../types';

export const graphApi = {
  /**
   * GET /api/v1/graph/entity/{value}
   */
  getEntity: (value: string): Promise<GraphNode> => {
    return apiFetch<GraphNode>(`/api/v1/graph/entity/${encodeURIComponent(value)}`);
  },

  /**
   * GET /api/v1/graph/entity/{value}/neighbors
   */
  getNeighbors: (value: string): Promise<GraphNeighborsResponse> => {
    return apiFetch<GraphNeighborsResponse>(
      `/api/v1/graph/entity/${encodeURIComponent(value)}/neighbors`
    );
  },

  /**
   * GET /api/v1/graph/entity/{value}/incidents
   */
  getRelatedIncidents: (value: string): Promise<RelatedIncidentsResponse> => {
    return apiFetch<RelatedIncidentsResponse>(
      `/api/v1/graph/entity/${encodeURIComponent(value)}/incidents`
    );
  },

  /**
   * GET /api/v1/graph/entity/{value}/risk
   */
  getEntityRisk: (value: string): Promise<EntityRiskResponse> => {
    return apiFetch<EntityRiskResponse>(
      `/api/v1/graph/entity/${encodeURIComponent(value)}/risk`
    );
  },

  /**
   * GET /api/v1/graph/entity/{value}/ring
   */
  getFraudRing: (value: string): Promise<FraudRingResponse> => {
    return apiFetch<FraudRingResponse>(
      `/api/v1/graph/entity/${encodeURIComponent(value)}/ring`
    );
  },

  /**
   * GET /api/v1/graph/network/summary
   */
  getNetworkSummary: (): Promise<NetworkSummaryResponse> => {
    return apiFetch<NetworkSummaryResponse>('/api/v1/graph/network/summary');
  },

  /**
   * GET /api/v1/graph/network/top-risk?limit=10
   */
  getTopRiskEntities: (limit = 10): Promise<TopRiskEntityResponse[]> => {
    return apiFetch<TopRiskEntityResponse[]>(
      `/api/v1/graph/network/top-risk?limit=${limit}`
    );
  },

  /**
   * GET /api/v1/graph/path?source=&target=
   */
  getShortestPath: (source: string, target: string): Promise<PathResponse> => {
    return apiFetch<PathResponse>(
      `/api/v1/graph/path?source=${encodeURIComponent(source)}&target=${encodeURIComponent(target)}`
    );
  },

  /**
   * GET /api/v1/graph/entity/{value}/shared
   */
  getSharedEntity: (value: string): Promise<SharedEntityResponse> => {
    return apiFetch<SharedEntityResponse>(
      `/api/v1/graph/entity/${encodeURIComponent(value)}/shared`
    );
  },

  /**
   * GET /api/v1/graph/visualization/{node_id}?depth=2
   */
  getVisualization: (nodeId: string, depth = 2): Promise<GraphResponse> => {
    return apiFetch<GraphResponse>(
      `/api/v1/graph/visualization/${encodeURIComponent(nodeId)}?depth=${depth}`
    );
  },
};
